lectures/dynamic_programming/wald_friedman.md

```diff
@@ -360,162 +360,181 @@ Later, doing this will help us obey **the don't repeat yourself (DRY)** golden r
 end
-```
 
-Next we solve a problem by finding the $\alpha$, $\beta$ values for the decision rule
+function decision_rule(problem; tol = 1e-6, max_iter = 400, verbose = false)
+    values = value_iteration(problem; tol = tol, max_iter = max_iter)
+    vf = extrapolate(interpolate((problem.belief_grid,), values, Gridded(Linear())), Flat())
 
-```{code-cell} julia
-function decision_rule(d0, d1, L0, L1, c)
-    function cost(p, d0, d1, L0, L1, c)
-        if c < zero(c)
-            throw(ArgumentError("Cost must be non-negative"))
-        end
-        x0 = accept_x0(p, L0)
-        x1 = accept_x1(p, L1)
-        draw = draw_again(p, d0, d1, L0, L1, c, min(x0, x1))
-        output = min(abs(draw - x0), abs(draw - x1), abs(x1 - x0))
-        return output
+    (; belief_grid, L0, L1) = problem
+    actions = similar(belief_grid, Int)
+
+    for (i, p) in enumerate(belief_grid)
+        stop0 = accept_x0(p, L0)
+        stop1 = accept_x1(p, L1)
+        cont = continuation_value(problem, p, vf)
+        val = min(stop0, stop1, cont)
+        actions[i] = isapprox(val, stop0; atol = 1e-5, rtol = 0) ? 1 :
+                     isapprox(val, stop1; atol = 1e-5, rtol = 0) ? 2 : 3
     end
-    # Find the indifference points
-    roots = find_zeros(p -> cost(p, d0, d1, L0, L1, c), 0 + eps(), 1 - eps())
-    # Compute the choice at both sides
-    left = first.(choice.(roots .- eps(), d0, d1, L0, L1, c))
-    right = first.(choice.(roots .+ eps(), d0, d1, L0, L1, c))
-    # Find beta by checking for a permanent transition from the area accepting to
-    # x₁ to never again accepting x₁ at the various indifference points
-    # Find alpha by checking for a permanent transition from the area accepting of
-    # x₀ to never again accepting x₀ at the various indifference points
-    beta = findlast((left .== 2) .& (right .≠ 2)) |>
-           (x -> isa(x, Int) ? roots[x] : 0)
-    alpha = findfirst((left .≠ 1) .& (right .== 1)) |>
-            (x -> isa(x, Int) ? roots[x] : 1)
-    if beta < alpha
-        @printf("Accept x1 if p <= %.2f\nContinue to draw if %.2f <= p <= %.2f
-                \nAccept x0 if p >= %.2f", beta, beta, alpha, alpha)
-    else
-        x0 = accept_x0(beta, L0)
-        x1 = accept_x1(beta, L1)
-        draw = draw_again(beta, d0, d1, L0, L1, c, min(x0, x1))
-        if draw == min(x0, x1, draw)
-            @printf("Accept x1 if p <= %.2f\nContinue to draw if %.2f <= p <= %.2f
-                    \nAccept x0 if p >= %.2f", beta, beta, alpha, alpha)
-        else
-            @printf("Accept x1 if p <= %.2f\nAccept x0 if p >= %.2f", beta,
-                    alpha)
-        end
+
+    beta_idx = findlast(actions .== 2)
+    alpha_idx = findfirst(actions .== 1)
+    beta = isnothing(beta_idx) ? belief_grid[1] : belief_grid[beta_idx]
+    alpha = isnothing(alpha_idx) ? belief_grid[end] : belief_grid[alpha_idx]
+
+    if verbose
+        @printf("Accept x1 if p <= %.3f\nContinue to draw if %.3f <= p <= %.3f\nAccept x0 if p >= %.3f",
```

Comment:
Remove the printf and the PrintF include at the top.  In Julia you generally should just use print and embedded code.  e.g. something like 

`println("Accept x1 if p <= {beta}...")` etc.  Since this is just debugging code it is OK if it doesn't format perfectly.

If you really want to round, people do `println("Accept x1 if p <= {round(beta,digits=2)}...")` or something like that.